packages/salesforcedx-vscode-apex/src/commands/apexActionController.ts

```diff
@@ -132,4 +132,173 @@ export class ApexActionController {
+      { modal: true },
+      nls.localize('overwrite'),
+      nls.localize('cancel')
+    );
+    return response === nls.localize('overwrite');
+  };
+
+  private getFolderForArtifact = async (): Promise<string | undefined> => {
+    const registryAccess = new RegistryAccess();
+
+    const defaultESRFolder = path.join(
+      workspaceUtils.getRootWorkspacePath(),
+      'force-app',
+      'main',
+      'default',
+      registryAccess.getTypeByName('ExternalServiceRegistration').directoryName
+    );
+    const folderUri = await vscode.window.showInputBox({
+      prompt: nls.localize('enter_esr_path'),
+      value: defaultESRFolder
+    });
+
+    return folderUri ? path.resolve(folderUri) : undefined;
+  };
+  private showNamedCredentialsQuickPick = async (): Promise<string | undefined> => {
+    let namedCredentials;
+    let selectedNamedCredential: string | undefined;
+    let finalNamedCredential: string | undefined;
+    try {
+      const rawQueryData = await (
+        await workspaceContext.getConnection()
+      ).query('SELECT MasterLabel FROM NamedCredential');
+      namedCredentials = {
+        ...rawQueryData,
+        records: this.flattenQueryRecords(rawQueryData.records)
+      };
+    } catch (parseError) {
+      throw new Error(nls.localize('error_parsing_nc'));
+    }
+
+    if (namedCredentials) {
+      const quickPicks = namedCredentials.records.map(nc => nc.MasterLabel as string);
+      quickPicks.push(nls.localize('enter_new_nc'));
+      selectedNamedCredential = await vscode.window.showQuickPick(quickPicks, {
+        placeHolder: nls.localize('select_named_credential')
+      });
+    }
+
+    if (!selectedNamedCredential || selectedNamedCredential === nls.localize('enter_new_nc')) {
+      finalNamedCredential = await vscode.window.showInputBox({
+        prompt: nls.localize('enter_nc_name')
+      });
+    } else {
+      finalNamedCredential = selectedNamedCredential;
+    }
+    return finalNamedCredential;
+  };
+
+  private flattenQueryRecords(rawQueryRecords: JsonMap[]) {
+    // filter out the attributes key
+    return rawQueryRecords.map(({ attributes, ...cleanRecords }) => cleanRecords);
+  }
+
+  private buildESRXml = (
+    existingContent: string | undefined,
+    fullPath: string,
+    namedCredential: string | undefined,
+    oasSpec: string
+  ) => {
+    const baseName = path.basename(fullPath).split('.')[0];
+    const safeOasSpec = oasSpec.replaceAll('"', '&apos;');
+
+    if (existingContent) {
+      const doc = create(existingContent);
+
+      // Correctly update the <schema> tag content
+      const schemaElement = doc.find((node: any) => node.node.nodeName === 'schema');
+      if (schemaElement) {
+        schemaElement.txt(safeOasSpec);
+      } else {
+        throw new Error('The <schema> element was not found in the provided XML.');
+      }
+
+      return doc.end({ prettyPrint: true });
+    }
+
+    const newDoc = create({ version: '1.0', encoding: 'UTF-8' })
```

Comment:
As we discussed yesterday, the `xmlbuilder2` module is fairly large, > 1MB. What other features does this offer aside from using it to build the xml doc for ESR?

Reply:
@CristiCanizales let leave it as is for now. Maybe revisit the xml building in Feb.